---
title: "*Hamadryas*  Geomorph"
author: "C. A. Hamm"
date: "March 12, 2015"
output: html_document
---

This is an R Markdown document version of the code for the Hamm, Penz and DeVries (submitted) manuscript:

**Wing shape evolution in *Hamadryas* butterflies corresponds to vertical microhabitat use and species range size**

```{r load package}
library("geomorph") #v2.1.4
library("phytools") #v0.4-31, loads ape v3.2 
library("geiger") #v2.0.3

# save(list = ls(), file = "Hama_data_1.R")
# load("Hama_data_1.R")
# packageDescription("ape")$Version
```

```{r read in tree data}
Hama <- read.nexus("../Tree_data/Hamydryas_ml.tre")
plot(Hama)
is.ultrametric(Hama)

# make tree ultrametric, lamdba is the smoothing parameter. 
lam <- 10^(-1:6)
cv <- sapply(lam, function(x) sum(attr(chronopl(Hama, lambda = x, 
  CV = TRUE), "D2")))
plot(lam, cv, pch = 19, ylab = "cross-validation score", 
	xlab = expression(paste(lambda)), las = 1, cex = 1.5) # lowest CV 

Hama2 <- chronopl(phy = Hama, lambda = 0.1, CV = TRUE, eval.max = 1e3, 
	iter.max = 1e4)
is.ultrametric(Hama2)
```

Now to read in morphometric data for **forewing** and conduct from preliminary steps

```{r read in forewing data}
fore <- readland.tps("dorsal_data/Hama_dorsal_aligned.tps",	specID = "ID")
str(fore)
dim(fore) # 50 landmarks with X & Y coordinates, for 17 samples

fore_gpa <- gpagen(A = fore, Proj = TRUE, ProcD = TRUE, 
	ShowPlot = TRUE) 
	# used Procrustes distance for sliding in TPSrelW
fw <- two.d.array(fore_gpa$coords)
match(Hama2$tip.label, row.names(fw)) # confirm names in tree and data match

# Because I constructed the sliders in a different program the centroid sizes calculated by gpagen are not correct. Here I import them manually (lame, I know). 
fore_cs <- c(3.33294570842831e+3, 3.16568469860016e+3, 
	3.39261539540194e+3, 3.78430193381669e+3, 3.48021469444961e+3, 
	3.33777956879712e+3, 2.68538836886851e+3,3.17658593061965e+3, 
	3.39194656501948e+3, 3.61375985829088e+3, 3.41355030541168e+3, 
	2.85453923363221e+3, 4.05270026183310e+3, 3.10518493006033e+3, 
	3.01038815277082e+3, 3.56187305827559e+3, 3.64253669986741e+3) 
fw_cs <- matrix(fore_cs, dimnames = list(names(fore_gpa$Csize)))
```

Now to read in data for **hind wing**

```{r hind wing data }
hind <- readland.tps("ventral_data/Hama_hind_struc.tps", specID = "ID")
str(hind)
dim(hind) # 12 landmarks with X & Y for 17 individuals

hind_gpa <- gpagen(A = hind, Proj = TRUE, ProcD = TRUE, ShowPlot = TRUE)
hw <- two.d.array(hind_gpa$coords)
match(Hama2$tip.label, row.names(hw))
match(row.names(fw), row.names(hw))

hw_cs <- matrix(hind_gpa$Csize, dimnames = list(names(hind_gpa$Csize)))
```

With all data read in and procrustes aligned we will test the shape data for phylogenetic signal

```{r phylo sig}
# forewing
(fw_ps <- physignal(phy = Hama2, A = fw, iter = 1e4, method = "Kmult")) # fore-wing shape is not randomly distributed through the tree

#centroid size
(fw_cs_ps <- physignal(phy = Hama2, A = fw_cs, iter = 1e4, 
	method = "Kmult")) # fore-wing centroid size is not randomly distributed through the tree

# phylogenetic signal
(hw_ps <- physignal(phy = Hama2, A = hw, iter = 1e4, method = "Kmult")) # hw shape is not randomly distributed through the tree

(hw_cs_ps <- physignal(phy = Hama2, A = hw_cs, iter = 1e4, 
  method = "Kmult")) # hindwing centroid size (as a proxy for size) is randomly distributed through the phylogeny, though shape is not.
```

Now to investigate our data a bit, first by performing a **PCA** on shape data

```{r PCA}
plotTangentSpace(A = fore_gpa$coords, label = TRUE, warpgrids = TRUE, 
	verbose = FALSE) # PC1 46%, PC2 24%

plotTangentSpace(A = hind_gpa$coords, label = TRUE) # PC1 (54%), PC2 (16%)
```

Now the real comparative stuff. Note, we will need to prune the tree for habitat data because we lack data for two *Hamadryas* species.

```{r comparative methods}
cu <- c(NA, 0, 0.5, 0.5, 0, 0.5, 0, 0, 0, 0.5, 0.5, 0, 0, 0 , NA, 1, 1) #0 = understory, 0.5 = mixed, 1 = canopy
cu <- matrix(cu, dimnames = list(names(hind_gpa$Csize)))
cu <- cu[Hama2$tip.label, ]
cu1 <- cu[-c(3, 9)]
cu1

Hama.pruned <- treedata(phy = Hama2, data = cu1, sort = TRUE, 
	warnings = TRUE) # removes alicia and julitta from data set (we have no ecological information for these)
```

We ask, accounting for phlyogeny, are **forewing** "integrated"? Then, are landmarks within wings correlated with one another? 

**We only want to compare "like" landmarks, such as type I only comapred to type I, and semi-landmarks compared only to semi-landmarks**


```{r forewing phylo.pls}
(F_H_pls <- phylo.pls(A1 = fore_gpa$coords, A2 = hind_gpa$coords, 
  phy = Hama2, warpgrids = TRUE, iter = 1e4)) # even when correcting for phylogeny there is no association

# examine correlations of characters within wings as well
plot(fore_gpa$coords[, 1, ], fore_gpa$coords[, 2, ], pch = 19, xlim = c(-0.21, 0.21), ylim = c(-0.1, 0.16))
# vein points
points(fore_gpa$coords[1:6, 1, ], fore_gpa$coords[1:6, 2, ], pch = 19, col = "red")
# anterior edge points
points(fore_gpa$coords[7:33, 1, ], fore_gpa$coords[7:33, 2, ], pch = 19, col = "Dodgerblue")
# posterior edge points
points(fore_gpa$coords[34:50, 1, ], fore_gpa$coords[34:50, 2, ], pch = 19, col = "dark green")

# compare leading edge and trailing edge of the wing
(LE_TE_pls <- phylo.pls(A1 = fore_gpa$coords[7:33, , ], 
  A2 = fore_gpa$coords[34:50, , ], phy = Hama2, warpgrids = TRUE, 
  iter = 1e4))
```

Now the same tests for hind wing

```{r hind wing phylo.pls}
plot(hind_gpa$coords[, 1, ], hind_gpa$coords[, 2, ], pch = 19, xlim = c(-0.4, 0.4), ylim = c(-0.3, 0.3))
points(hind_gpa$coords[1:6, 1, ], hind_gpa$coords[1:6, 2, ], col = "dodgerblue", pch = 19)
points(hind_gpa$coords[7:12, 1, ], hind_gpa$coords[7:12, 2, ], col = "red", pch = 19)

(HW_pls <- phylo.pls(A1 = hind_gpa$coords[1:6, , ], A2 = 
	hind_gpa$coords[7:12, , ],	phy = Hama2, warpgrids = TRUE, 
	iter = 1e4))
```





```{r compare rate of phenotypic change by habitat while accounting for phylogeny}
# compare canopy and understory rates
cuf <- as.factor(cu)

# forewing rates
(fw_habitat_rate <- compare.evol.rates(phy = Hama2, 
	A = fore_gpa$coords, gp = cuf, iter = 1e4))

# hindwing rates
(hw_habitat_rate <- compare.evol.rates(phy = Hama2, 
	A = hind_gpa$coords, gp = cuf, iter = 1e4))
```
Now to compare rates of phenotypic change by range size while accounting for phylogeny

```{r range rates}  
# Small <= 50000; 50001 > Medium < 100000; 100001 > Large 
ranges <- read.csv("Hamadryas_range.csv", header = TRUE, row.names = 1)
ranges$Range <- as.factor(ranges$Range)
str(ranges)

H.range <- matrix(ranges$Range, dimnames = list(row.names(ranges)))
H.range <- as.factor(H.range)
names(H.range) <- row.names(ranges)
H.range
H.range <- H.range[Hama2$tip.label]

(fw_range_rate <- compare.evol.rates(phy = Hama2, A = fore_gpa$coords, 
	gp = H.range, iter = 1e4)) #runs in support to the notion that reduced range species have accelerated rates of Evolution

(hw_range_rate <- compare.evol.rates(phy = Hama2, A = hind_gpa$coords, 
	gp = H.range, iter = 1e4)) # for hindwing the small range has the lowest rate. 
```

Now to make the data figure. 

```{r Figure 2}
Hama.range <- c("Large", "Small", "Small", "Medium", "Large", "Large", "Medium", "Medium", "Small", "Large", "Large", "Large", "Large", "Large", "Large", "Small", "Large")
Hama.range <- matrix(Hama.range, dimnames = list(Hama2$tip.label))

cu_color <- cu
cu_color[cu == 0] <- "brown"
cu_color[cu == 0.5] <- "orange"
cu_color[cu == 1] <- "dark green"

range.color <- as.integer(H.range)
range.color[range.color == "3"] <- "red"
range.color[range.color == "2"] <- "green"
range.color[range.color == "1"] <- "dodgerblue"

# pdf(file = "Hamadryas_phylo2.pdf")
plot.phylo(Hama2, font = 3, label.offset = 0.03, edge.width = 3)
points(rep(0.935, length(Hama2$tip.label)), 1 : length(Hama2$tip.label), pch = 19, col = range.color, cex = 1.5)
points(rep(0.98, length(Hama2$tip.label)), 1 : length(Hama2$tip.label), pch = 19, col = cu_color, cex = 1.5)
legend("topleft", legend = c("Range", "Small", "Medium", "Large", "Habitat", "Ground", "Canopy", "Mixed"), ncol = 2, pch = 19, col = c(0, "red", "green", "dodgerblue", 0, "darkred", "darkgreen", "orange"), bty = "n", pt.cex = 1.5)
# dev.off()
```

Figure 3 was made using hacked code from the **geomorph** package

Here is the code to make the supplemental material figure. Here we ask about ancestral traits for habitat use. 

```{r supp mat}

# pdf(file = "Hama_FastAnc.pdf", bg = "white")
contMap(Hama.pruned$phy, x = cu1, res = 1000)
# dev.off()

#fan plot used for a pretty cover photo
# pdf(file = "Hama_fan.pdf", bg = "white")
# fan <- contMap(Hama.pruned$phy, x = cu1, res = 1000, type = "fan", legend = FALSE)
# dev.off()
```